components/raftstore/src/store/peer_storage.rs

```diff
@@ -833,6 +837,19 @@ where
     #[inline]
     pub fn set_applied_state(&mut self, apply_state: RaftApplyState) {
         self.apply_state = apply_state;
+        let mut snap_state = self.snap_state.borrow_mut();
+        if let SnapState::Generating {
+            ref canceled,
+            ref index,
+            ..
+        } = *snap_state
+        {
+            let snap_index = index.load(Ordering::SeqCst);
+            if snap_index > 0 && self.apply_state.get_truncated_state().index > snap_index {
```

Comment:
How about exposing a method, and allow called from outside? So it can be called in various `on_ready_xxx`.